File: LeetCode-Python/combination_sum_II.py
# Given a collection of candidate numbers (C) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.

# Each number in C may only be used once in the combination.

# Note:
# All numbers (including target) will be positive integers.
# Elements in a combination (a1, a2, ... , ak) must be in non-descending order. (ie, a1 <= a2 <= ... ≤ ak).
# The solution set must not contain duplicate combinations.
# For example, given candidate set 10,1,2,7,6,1,5 and target 8, 
# A solution set is: 
# [1, 7] 
# [1, 2, 5] 
# [2, 6] 
# [1, 1, 6] 

class Solution:
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def combinationSum2_1_dfs(self, candidates, target):

        def dfs(candidates, target, start, valuelist):
            if target == 0:
                if valuelist not in res: res.append(valuelist)
                return
            for i in xrange(start, len(candidates)):
                if target < candidates[i]:
                    return
                dfs(candidates, target - candidates[i], i + 1, valuelist + [candidates[i]])

        candidates.sort()
        res = []
        dfs(candidates, target, 0, [])
        return res

File: LintCodeJava/CombinationSumII.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSumII {
    /**
     * @param candidates: Given the candidate numbers
     * @param target: Given the target number
     * @return: All the combinations that sum to target
     */

    public static List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        if (candidates == null) {
            return result;
        }

        ArrayList<Integer> list = new ArrayList<Integer>();
        Arrays.sort(candidates);
        helper(candidates, target, list, 0, result);

        return result;
    }
    private static void helper(int[] candidates, int target, ArrayList<Integer> list, int pos, List<List<Integer>> result) {
        if (target == 0) {
            result.add(new ArrayList<Integer>(list));
            return;
        }

        int prev = -1;
        for (int i = pos; i < candidates.length; i++) {
            if (candidates[i] > target) {
                break;
            }

            // Skip duplicate numbers on same recursion tree level
            if (prev != -1 && prev == candidates[i]) {
                continue;
            }

            list.add(candidates[i]);
            helper(candidates, target - candidates[i], list, i + 1, result);
            list.remove(list.size() - 1);

            prev = candidates[i];
        }
    }
}

